### ISSUE TITLE
Impossible to use a Vagrantfile created by the same Terraform process?

### ISSUE THREAD

**Issue by petr-motejlek** [bmatcuk/terraform-provider-vagrant#10]
Hey.

It seems to me, it is rather difficult to get the `vagrant_vm` resource to properly wait for a `local_file` resource to actually create the `Vagrantfile`. When I use `dirname(local_file.Vagrantfile.filename)`, I always get a complaint about `stat` not being able to find the Vagrantfile. Go figure, the `local_file.Vagrantfile` resource hasn't even ran yet, and the Vagrant provider seems to try to access it ...

I tried using `depends_on = [local_file.Vagrantfile]` inside the `vagrant_vm` resource, but that's not helping at all. The provider seems to want to `stat` the Vagrantfile very early on.

The only way I found how to make sure the `vagrant_vm` resource will actually wait for a `local_file` to get created prior to trying to `stat` it was via a `data "null_data_source" "Vagrantfile"` block that would use the `local_file.Vagrantfile.filename`, and then use  `depends_on = [data.null_data_source.Vagrantfile]` inside the `vagrant_vm` block. This forces the Vagrant provider to wait for the file to get created before it `stat`s it, however, I am always greeted by a deprecation warning from Terraform, that using a `null_data_source` is deprecated and that `locals` does the same job -- well, it doesn't. At least not with the Vagrant provider :).

---

Am I doing it wrong, or is the Vagrant provider really just too eager to `stat` the Vagrantfile too early? What would be the correct way to use `resource "local_file" "Vagrantfile"` to generate a Vagrantfile and then feed it into the `vagrant_vm` resource?

**Comment by petr-motejlek** [comment]
Actually ... I was wrong when I said that the `null_data_source` can be used :). I kinda messed up, and left the `Vagrantfile` created by an earlier `terraform apply` there. Hence, `vagrant_vm` saw it, and was happy :D.

So ... it seems like the only viable workaround I found so far was to `touch` the `Vagrantfile` before running `terraform apply` (or `terraform destroy`). It's not really a good one, as `local_file` will remove the file when it gets destroyed. So ... you'd have to somehow magically always touch it.

**Comment by bmatcuk** [comment]
Hmm, you could try passing in either `local_file.Vagrantfile.filename` or `local_file.Vagrantfile.id` as an environment variable. That might be enough to trick terraform into waiting for the file to write:

```terraform
resource "vagrant_vm" "my_vm" {
  env = {
    VAGRANTFILE_CHECKSUM = local_file.Vagrantfile.id
  }
}
```

The name of the environment variable doesn't matter.

Now that I think of it, you may want to use `local_file.Vagrantfile.id` because, [looking at the code for local_file](https://github.com/hashicorp/terraform-provider-local/blob/9b4bb59773b92d3e019b0b4ca156ae84d9eaefde/internal/provider/resource_local_file.go#L142), the id updates when the file's content updates. So, this _should_ cause terraform to re-run vagrant if the file's content changes.

**Comment by petr-motejlek** [comment]
I tried multiple different alternatives, and nothing stuck. It seems to me like the `vagrant_vm` resources is just too eager to try to `stat` the `Vagrantfile`. Why is it even trying to to that during the "refresh" phase?

```
resource "vagrant_vm" "nodes" {
  vagrantfile_dir = dirname(local_file.Vagrantfile.filename)
  env = {
    VAGRANT_EXPERIMENTAL = "disks",
    CAFILE_HASH          = md5(data.local_file.ca_public_key.content)
    VAGRANTFILE_HASH     = md5(local_file.Vagrantfile.content)
    VAGRANTFILE_ID       = local_file.Vagrantfile.id
    PROVISIONSHFILE_HASH = md5(local_file.provision_sh.content)
  }
  get_ports = true
}
```

**Comment by petr-motejlek** [comment]
Yeah, that might be a good lead with the "refresh" phase. You cannot actually even do `terraform refresh`, unless you `touch` the `Vagrantfile` first in a setup with `local_file.Vagrantfile` :D I realize it's a chicken-n-egg situation, but shouldn't this be supported too somehow?

**Comment by bmatcuk** [comment]
Unfortunately, that's kind of just how terraform "works"... A provider can only define the four basic CRUD methods, so there's no opportunity to, say, do something different during the refresh phase. I'm not sure how I can solve this problem. I spent some time yesterday trying to search for a solution, but I only found people complaining about similar issues with other providers.

If you're aware of a provider that solves this issue, let me know - I could take a look at how they did it.

**Comment by petr-motejlek** [comment]
I'll take a look. I actually never wrote nor saw any provider code just yet :). But I would think there must be a way for you to really begin reading the file when you are either doing any of the actions, and not just when refreshing. 

For now, I have manged to find a workaround that's still not perfect, but is somewhat OK.

You do have to keep an empty `Vagrantfile` around all the time, though. This Terraform block just makes sure that it always exists, even when Terraform's `local_file` would remove it. It also probably only works on POSIX systems.

The `create` lifecycle command could probably be blank, as we really only need to make sure that `delete` leaves some file in there.

```
locals {
  touch_Vagrantfile = "echo > ${local.Vagrantfile_path}"
  Vagrantfile_path  = abspath("${path.module}/Vagrantfile")
}

resource "shell_script" "Vagrantfile" {
  # Dummy resource to make sure that "some" Vagrantfile is always present
  # because of https://github.com/bmatcuk/terraform-provider-vagrant/issues/10

   lifecycle_commands {
     create = local.touch_Vagrantfile
     delete = local.touch_Vagrantfile
   }
}
      
resource "local_file" "Vagrantfile" {
  depends_on = [shell_script.Vagrantfile]

  filename = local.Vagrantfile_path
  content  = "..."
}

resource "vagrant_vm" "nodes" {
  vagrantfile_dir = dirname(local_file.Vagrantfile.filename)
  env = {
    VAGRANT_EXPERIMENTAL = "disks",
    VAGRANTFILE_HASH     = md5(local_file.Vagrantfile.content)
  }
  get_ports = true
}
```

The reason for this weird wonky all-over-the-place shell-scripting is because `local_file` will upon destroy remove the `Vagrantfile` from disk. Then, any subsequent `terraform apply` or `terraform plan` or `terraform refresh` will blow up, because `vagrant_vm` needs "a Vagrantfile" there all the time. Hence, I made the `local_file.Vagrantfile` depend on a dummy resource that is always deleted "AFTER" the `Vagrantfile`, and hence can recreate it empty.

**Comment by vfalfushinsky** [comment]
I've managed to use a Vagrant file that is created by the same Terraform process, but if it is located in the same directory. Using different directory for it and using `vagrantfile_dir` parameter causes an error which occurs on plan:
```
Error: stat /Users/test/sources/terraform-vagrant-vm/_examples/latest/my-vm-001/Vagrantfile: no such file or directory
```

**Comment by bmatcuk** [comment]
You're probably running into the same problem as petr-motejlek above: the file needs to exist. You could create an empty Vagrantfile.

**Comment by vfalfushinsky** [comment]
It looks like it, but I also tried custom Vagrantfile without custom directory and it worked fine.
